egs/kinect-wsj/DeepClustering/train.py

```diff
@@ -0,0 +1,207 @@
+from asteroid.losses import deep_clustering_loss
+from asteroid.filterbanks.transforms import take_mag, ebased_vad
+
+from asteroid.data.kinect_wsj import make_dataloaders
+from model import make_model_and_optimizer
+
+EPS = 1e-8
+
+parser = argparse.ArgumentParser()
+parser.add_argument('--exp_dir', default='exp/tmp',
+                    help='Full path to save best validation model')
+
+
+def main(conf):
+    exp_dir = conf['main_args']['exp_dir']
+    # Define Dataloader
+    train_loader, val_loader = make_dataloaders(**conf['data'],
+                                                **conf['training'])
+    conf['masknet'].update({'n_src': conf['data']['n_src']})
+    # Define model, optimizer + scheduler
+    model, optimizer = make_model_and_optimizer(conf)
+    scheduler = None
+    if conf['training']['half_lr']:
+        scheduler = ReduceLROnPlateau(optimizer=optimizer, factor=0.5,
+                                      patience=5)
+
+    # Save config
+    os.makedirs(exp_dir, exist_ok=True)
+    conf_path = os.path.join(exp_dir, 'conf.yml')
+    with open(conf_path, 'w') as outfile:
+        yaml.safe_dump(conf, outfile)
+
+    # Define loss function
+    loss_func = ChimeraLoss(alpha=conf['training']['loss_alpha'])
+    # Put together in System
+    system = ChimeraSystem(model=model, loss_func=loss_func,
+                           optimizer=optimizer, train_loader=train_loader,
+                           val_loader=val_loader, scheduler=scheduler,
+                           config=conf)
+
+    # Callbacks
+    checkpoint_dir = os.path.join(exp_dir, 'checkpoints/')
+    checkpoint = ModelCheckpoint(checkpoint_dir, monitor='val_loss',
+                                 mode='min', save_top_k=5, verbose=1)
+    early_stopping = False
+    if conf['training']['early_stop']:
+        early_stopping = EarlyStopping(monitor='val_loss', patience=10,
+                                       verbose=1)
+    gpus=-1
+    # Don't ask GPU if they are not available.
+    if not torch.cuda.is_available():
+        print('No available GPU were found, set gpus to None')
+        gpus = None
+
+    # Train model
+    trainer = pl.Trainer(max_nb_epochs=conf['training']['epochs'],
+                         checkpoint_callback=checkpoint,
+                         early_stop_callback=early_stopping,
+                         default_save_path=exp_dir,
+                         gpus=gpus,
+                         distributed_backend='dp',
+                         train_percent_check=1.0,  # Useful for fast experiment
+                         gradient_clip_val=200,)
+    trainer.fit(system)
+
+    with open(os.path.join(exp_dir, "best_k_models.json"), "w") as f:
+        json.dump(checkpoint.best_k_models, f, indent=0)
+    # Save last model for convenience
+    torch.save(system.model.state_dict(),
+               os.path.join(exp_dir, 'checkpoints/final.pth'))
+
+
+    # TODO:Should ideally be inherited from wsj0-mix
```

Comment:
I'd love to reduce the copy pasting from recipes to recipes but apart from adding this things into the source code, I didn't find a good solution yet. Any good idea welcome !

Reply:
Is Deep clustering the only recipe which requires a custom System class?

Reply:
No, also in DNS challenge and the two step recipe

Reply:
DC models can probably go into blocks.
System callbacks can handle common_steps I think. Will have to analyse a bit more to see if it gets complicated.